CMakeLists.txt

```diff
@@ -239,6 +250,80 @@ endif()
 
 include(cmake/optional.cmake)
 
+set(hardened_cxx_flags "")
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  # All versions of gcc that we commonly use for building are subject to bug
+  # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90348. To work around that, set
+  # -fstack-reuse=none for all gcc builds.
+  try_append_cxx_flags(hardened_cxx_flags "-fstack-reuse=none")
+endif()
+if(HARDENING)
+  if(MSVC)
+    try_append_cxx_flags(core "/DYNAMICBASE")
+    try_append_cxx_flags(core "/HIGHENTROPYVA")
+    try_append_cxx_flags(core "/NXCOMPAT")
+  else()
+    try_append_cxx_flags(hardened_cxx_flags "-Wstack-protector")
+    try_append_cxx_flags(hardened_cxx_flags "-fstack-protector-all")
+    try_append_cxx_flags(hardened_cxx_flags "-fcf-protection=full")
+    if(MINGW)
+      link_libraries(ssp)
+    else()
+      # stack-clash-protection doesn't currently work, and likely should just be skipped for Windows.
+      # See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90458 for more details.
+      try_append_cxx_flags(hardened_cxx_flags "-fstack-clash-protection")
+    endif()
+
+    try_append_linker_flag(core "-Wl,--enable-reloc-section")
```

Comment:
I guess same question here as above? Can we not have `hardend_ldflags`?

Reply:
I'll answer this comment shortly.

Reply:
> I guess same question here as above? Can we not have `hardend_ldflags`?

In the current state of CMake-based build system, the `try_append_cxx_flags` and `try_append_linker_flag` functions accept different types of the first argument.

The `try_append_cxx_flags` function works with a variable, whose value presents a string of concatenated flags. This perfectly follows the Make's approach and is still needed for modifying `CMAKE_CXX_FLAGS_<CONFIG>` configuration-dependant flags.

The `try_append_linker_flag` function uses the modern CMake's approach working with targets and their properties.

I really do prefer the latter. Besides, what is the point of introducing intermediate variables like `something_ldflags`?

However, for the sake of consistency, `try_append_cxx_flags` should also have an option to work with targets. For example, its signature could allow the usage as follows:
```cmake
try_append_cxx_flags("-Werror" VAR error_cxx_flags)
```
or
```cmake
try_append_cxx_flags("-Werror" TARGET core)
```

Reply:
Would it be possible to get both by appending flags to a virtual lib, then "linking" in that lib to get its flags?

Reply:
> Would it be possible to get both by appending flags to a virtual hardening lib, then "linking" in that lib to get its flags?

Just to clarify your point, what kind of function signature do you imagine?

Reply:
I'm suggesting nested interfaces (if possible). something like:
```CMake
add_library(hardening INTERFACE)
if(HARDENING)
  try_append_linker_flag(hardening "-Wl,--enable-reloc-section")
endif()
target_link_libraries(core PRIVATE hardening)
```

Reply:
> I really do prefer the latter. Besides, what is the point of introducing intermediate variables like something_ldflags?

I assume @fanquake's issue here, like mine, is that if we add these to `core` directly, we have no way of referring to the hardening options alone?

If I understand, you're suggesting that we add:
```CMake
try_append_cxx_flags("foo" TARGET bar)
```

If that understanding is correct, then yes, I agree we should add that construction. I believe it would allow us to do:

```CMake
add_library(hardening INTERFACE)
if(HARDENING)
  try_append_cxx_flags("-fstack-clash-protection" TARGET hardening)
  try_append_linker_flag(hardening "-Wl,--enable-reloc-section")
endif()
target_link_libraries(core PRIVATE hardening)
```

And it should just work, no?

Reply:
@theuni 

Thanks! That looks like a solution :)